engine/src/main/java/org/terasology/rendering/dag/BloomPassesNode.java

```diff
@@ -0,0 +1,149 @@
+import org.terasology.config.Config;
+import org.terasology.config.RenderingConfig;
+import org.terasology.monitoring.PerformanceMonitor;
+import org.terasology.registry.In;
+import org.terasology.rendering.assets.material.Material;
+import org.terasology.rendering.nui.properties.Range;
+import org.terasology.rendering.opengl.FBO;
+import org.terasology.rendering.opengl.FrameBuffersManager;
+import org.terasology.rendering.world.WorldRenderer;
+import static org.lwjgl.opengl.GL11.GL_COLOR_BUFFER_BIT;
+import static org.lwjgl.opengl.GL11.GL_DEPTH_BUFFER_BIT;
+import static org.lwjgl.opengl.GL11.glClear;
+import static org.terasology.rendering.opengl.OpenGLUtils.bindDisplay;
+import static org.terasology.rendering.opengl.OpenGLUtils.setViewportToSizeOf;
+import static org.terasology.rendering.opengl.OpenGLUtils.renderFullscreenQuad;
+
+/**
+ * TODO: Add diagram of this node
+ */
+public class BloomPassesNode implements Node {
+
+    @Range(min = 0.0f, max = 5.0f)
+    private float bloomHighPassThreshold = 0.05f;
+
+    @Range(min = 0.0f, max = 32.0f)
+    private float bloomBlurRadius = 12.0f;
+
+    @In
+    private Config config;
+
+    @In
+    private FrameBuffersManager frameBuffersManager;
+
+    @In
+    private WorldRenderer worldRenderer;
+
+    private RenderingConfig renderingConfig;
+    private Material highPass;
+    private Material blur;
+    private FBO sceneOpaque;
+    private FBO sceneBloom0;
+    private FBO sceneBloom1;
+    private FBO sceneBloom2;
+    private FBO sceneHighPass;
+
+    @Override
+    public void initialise() {
+        renderingConfig = config.getRendering();
+        blur = worldRenderer.getMaterial("engine:prog.blur");
+        highPass = worldRenderer.getMaterial("engine:prog.highp"); // TODO: rename shader to highPass
+    }
+
+    /**
+     * If bloom is enabled via the rendering settings, this method generates the images needed
+     * for the bloom shader effect and stores them in their own frame buffers.
+     * <p>
+     * This effects renders adds fringes (or "feathers") of light to areas of intense brightness.
+     * This in turn give the impression of those areas partially overwhelming the camera or the eye.
+     * <p>
+     * For more information see: http://en.wikipedia.org/wiki/Bloom_(shader_effect)
+     */
+    @Override
+    public void process() {
+        if (renderingConfig.isBloom()) {
+            PerformanceMonitor.startActivity("rendering/bloompasses");
+
+            sceneBloom0 = frameBuffersManager.getFBO("sceneBloom0");
+            sceneBloom1 = frameBuffersManager.getFBO("sceneBloom1");
+            sceneBloom2 = frameBuffersManager.getFBO("sceneBloom2");
+            sceneHighPass = frameBuffersManager.getFBO("sceneHighPass");
+            sceneOpaque = frameBuffersManager.getFBO("sceneOpaque");
+
+            generateHighPass();
+            generateBloom(sceneBloom0);
+            generateBloom(sceneBloom1);
+            generateBloom(sceneBloom2);
```

Comment:
Let's add to this block: // TODO: review - would it make sense to split these operations into one highpass node and three blur nodes with different parameters?

Interestingly, this hints to a bigger issue: if a number of "elementary" nodes are working closely together to achieve a more complex effect, how can they be "grouped" together? Perhaps this could be handled by considering some nodes as a graph within the graph, which is functionality we probably need anyway, i.e. to handle the eye-based parallel graphs for stereoscopic displays.